File: Halo4Api.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime;
using System.Text;
using System.Collections.Specialized;
using System.IO;
using System.Net;
using System.Text.RegularExpressions;
using System.Web;
using Newtonsoft.Json;

public class Halo4Api
{
    private class BetterWebRequest
    {
        public static HttpWebResponse Post(string url, string body) // Implies "POST"
        {
            HttpWebRequest request = (HttpWebRequest)HttpWebRequest.Create(url);
            request.AllowAutoRedirect = false;
            request.Method = "POST";
            request.CookieContainer = Halo4Api.Cookies;
            HttpWebResponse response;
            System.Text.ASCIIEncoding newBody = new System.Text.ASCIIEncoding();
            request.ContentLength = body.Length;
            request.ContentType = "application/x-www-form-urlencoded";

            StreamWriter writer = new StreamWriter(request.GetRequestStream());
            writer.Write(body);
            writer.Close();

            response = (HttpWebResponse)request.GetResponse();
            response.Close();

            return response;
        }
    }

    private class BetterWebResponse
    {
        public static string GetBody(HttpWebResponse response)
        {
            try {
                Stream dataStream = response.GetResponseStream();

                StreamReader reader = new StreamReader(dataStream);

                string body = reader.ReadToEnd();

                dataStream.Close();
                reader.Close();

                return body;
            }
            catch(Exception)
            {
                return null;
            }
        }
    }

    private class WaypointAuthTokenDocument
    {
        public string access_token { get; set; }
        public string AuthenticationToken { get; set; }
        public int expires_in { get; set; }
    }

    private class UserInformation
    {
        public string Gamertag { get; set; }
        public string AnalyticsToken { get; set; }
    }

    private class WaypointSpartanTokenDocument
    {
        public int ResponseCode { get; set; }
        public string SpartanToken { get; set; }
        private UserInformation UserInformation { get; set; }
    }

    // Enter an email or pass here to use the GetAPIKey() method without passing values
    private string microsoftEmail = "";
    private string microsoftPassword = "";

    private string msLogin = "https://login.live.com/login.srf?id=2";
    private string waypointGateway = "https://app.halowaypoint.com/oauth/signin?returnUrl=https%3A%2F%2Fapp.halowaypoint.com%2Fen-us";
    private string waypointRegisterUrl = "https://settings.svc.halowaypoint.com/RegisterClientService.svc/spartantoken/wlid?_={0}";

    // New URL's
    private string urlToScrape = "https://login.live.com/oauth20_authorize.srf?client_id=000000004C0BD2F1&scope=xbox.basic+xbox.offline_access&response_type=code&redirect_uri=https://www.halowaypoint.com:443/oauth/callback&state=https%3A%2F%2Fapp.halowaypoint.com%2Fen-us%2F&locale=en-US&display=touch";
    private string urlToPost = "https://login.live.com/ppsecure/post.srf?client_id=000000004C0BD2F1&scope=xbox.basic+xbox.offline_access&response_type=code&redirect_uri=https://www.halowaypoint.com:443/oauth/callback&state=https%3A%2F%2Fapp.halowaypoint.com%2Fen-us%2F&locale=en-US&display=touch&bk=1383096785";


    public static CookieContainer Cookies = new CookieContainer();

    private string GetMicrosoftLoginPPFT(int iterationCount = 0) // Going to scrape a page and return an array of strings containing essential values for the login transaction
    {
        HttpWebRequest request = (HttpWebRequest)HttpWebRequest.Create(urlToScrape);
        HttpWebResponse response;
        try
        {
            response = (HttpWebResponse)request.GetResponse();
        }
        catch(Exception)
        {
            response = null;
        }

        if (response == null && iterationCount < 10)
        {
            return GetMicrosoftLoginPPFT(iterationCount++);
        }
        else if(iterationCount == 10)
        {
            if (response != null)
            {
                response.Close();
            }
            throw new Exception("Failed to get login parameters after 10 attempts, Microsoft may be down.");
        }

        string responseBody = BetterWebResponse.GetBody(response);

        Regex cookieRegex = new Regex(@"MSPRequ=(.*?);.*,.*MSPOK=(.*?);");

        Match cookieMatch = cookieRegex.Match(response.Headers.Get("Set-Cookie"));

        Cookie mspRequest = new Cookie("MSPRequ", cookieMatch.Groups[1].Value, "/", "login.live.com");

        Cookie mspOk = new Cookie("MSPOK", cookieMatch.Groups[2].Value, cookieMatch.Groups[6].Value, "login.live.com");

        Halo4Api.Cookies.Add(mspRequest);
        Halo4Api.Cookies.Add(mspOk);

        response.Close();

        Regex regex = new Regex("name=\"PPFT\".*?value=\"(.*?)\"");

        Match match = regex.Match(responseBody);

        string PPFT;

        if (match.Success)
        {
            PPFT = match.Groups[1].Value;
        }
        else
        {
            throw new Exception("Failed to get login parameters. Scrape returned no results, Microsoft may have changed their code.");
        }

        return PPFT;
    }

    private string GetWaypointState(int iterationCount = 0)
    {
        HttpWebRequest request = (HttpWebRequest)HttpWebRequest.Create(waypointGateway);
        request.CookieContainer = Halo4Api.Cookies;
        HttpWebResponse response;
        try
        {
            response = (HttpWebResponse)request.GetResponse();
        }
        catch (Exception)
        {
            response = null;
        }

        if (response == null && iterationCount < 10)
        {
            return GetWaypointState(++iterationCount);
        }
        else if (iterationCount == 10)
        {
            throw new Exception("Failed to get login parameters after 10 attempts, Microsoft may be down.");
        }
        
        string waypointState;
        try
        {
            NameValueCollection queryString = HttpUtility.ParseQueryString(response.ResponseUri.Query);
            waypointState = queryString["state"];
        }
        catch (Exception)
        {
            return GetWaypointState(++iterationCount);
        }

        Halo4Api.Cookies.Add(response.Cookies);

        response.Close();

        return waypointState;
    }

    private string PerformWaypointLogin(string email, string pass)
    {
        string PPFT = GetMicrosoftLoginPPFT();
        string PPSX = "Pass";

        /*
           PPFT:[loginPPFT]
           login:[login]
           passwd:[passwd]
           LoginOptions:3
           NewUser:1
           PPSX:[loginPPSX]
           type:11
           i3:[rand]
           m1:1680
           m2:1050
           m3:0
           i12:1
           i17:0
           i18:__MobileLogin|1
         */
        string query = String.Format("PPFT={0}&login={1}&passwd={2}&LoginOptions=3&NewUser=1&PPSX={3}&type=11&i3={4}&m1=1680&m2=1050&m3=0&i12=1&i17=0&i18=__MobileLogin|1", PPFT, HttpUtility.UrlEncode(email), HttpUtility.UrlEncode(pass), PPSX, new Random().Next(15000, 50000).ToString());

        // Get the response of the query using the parameters we created
        HttpWebResponse response = BetterWebRequest.Post(urlToPost, query);

        // Add the response cookies to our container
        Halo4Api.Cookies.Add(response.Cookies);

        // Close the response object
        response.Close();

        // Return the header that we need
        return response.Headers.Get("Location");
    }

    private WaypointAuthTokenDocument GetWaypointAuthToken(string homepageUrl)
    {
        HttpWebRequest request = (HttpWebRequest)HttpWebRequest.Create(homepageUrl);
        request.CookieContainer = Halo4Api.Cookies;
        request.UserAgent = "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.17 (KHTML, like Gecko) Chrome/24.0.1312.52 Safari/537.17";

        WaypointAuthTokenDocument authToken;

        HttpWebResponse response;

        try
        {
            response = (HttpWebResponse)request.GetResponse();

            Regex regex = new Regex("var user.=.*?({.*?});", RegexOptions.Multiline);

            string body = BetterWebResponse.GetBody(response);

            Match match = regex.Match(body);

            if (!match.Success)
            {
                throw new Exception();
            }

            authToken = JsonConvert.DeserializeObject<WaypointAuthTokenDocument>(match.Groups[1].Value);

            Halo4Api.Cookies.Add(response.Cookies);
        }
        catch (Exception)
        {
            throw new Exception("Failed to get Waypoint auth tokens. This is likely a scraping error. Perhaps 343i changed their code layout.");
        }

        response.Close();

        return authToken;
    }

    private WaypointSpartanTokenDocument GetWaypointSpartanToken(WaypointAuthTokenDocument authToken)
    {
        HttpWebRequest request = (HttpWebRequest)HttpWebRequest.Create(String.Format(waypointRegisterUrl, (int)(DateTime.UtcNow - new DateTime(1970, 1, 1)).TotalSeconds));
        request.Accept = "application/json";
        request.Headers.Add("Origin", "https://app.halowaypoint.com");
        request.Headers.Add("X-343-Authorization-WLID", "v1=" + authToken.access_token);
        request.CookieContainer = Halo4Api.Cookies;

        HttpWebResponse response = (HttpWebResponse)request.GetResponse();

        WaypointSpartanTokenDocument spartanToken;

        try
        {
            spartanToken = JsonConvert.DeserializeObject<WaypointSpartanTokenDocument>(BetterWebResponse.GetBody(response));
        }
        catch (Exception)
        {
            throw new Exception("Could not parse Spartan token.");
        }

        Halo4Api.Cookies.Add(response.Cookies);

        return spartanToken;
    }

    private string GetWaypointWebAuthTokenAndHomepage(string callbackUrl)
    {
        HttpWebRequest request = (HttpWebRequest)HttpWebRequest.Create(callbackUrl);
        request.CookieContainer = Halo4Api.Cookies;
        request.AllowAutoRedirect = false;

        HttpWebResponse response = (HttpWebResponse)request.GetResponse();

        bool webAuth = false;

        foreach (Cookie cookie in response.Cookies)
        {
            if (cookie.Name == "WebAuth")
            {
                webAuth = true;
                break;
            }
        }

        if (webAuth == false)
        {
            throw new Exception("Could not get WebAuth token.");
        }

        Halo4Api.Cookies.Add(response.Cookies);

        response.Close();

        return response.Headers.Get("Location");
    }

    private string GetSpartanTokenViaMicrosoftAuth(string email, string pass)
    {
        string waypointCallbackUrl = PerformWaypointLogin(email, pass);
        string waypointHomepage = GetWaypointWebAuthTokenAndHomepage(waypointCallbackUrl); // Loads the WebAuth token into the cookie store
        WaypointAuthTokenDocument waypointAuthToken = GetWaypointAuthToken(waypointHomepage);
        WaypointSpartanTokenDocument waypointSpartanToken = GetWaypointSpartanToken(waypointAuthToken);

        return waypointSpartanToken.SpartanToken;
    }

    public string GetAPIKey()
    {
        try
        {
            return GetSpartanTokenViaMicrosoftAuth(microsoftEmail, microsoftPassword);
        }
        catch (Exception e)
        {
            throw new Exception(e.Message != null ? e.Message : "An unknown error occurred.");
        }
    }

    public string GetAPIKey(String email, String pass)
    {
        try
        {
            return GetSpartanTokenViaMicrosoftAuth(email, pass);
        }
        catch (Exception e)
        {
            throw new Exception(e.Message != null ? e.Message : "An unknown error occurred.");
        }
    }
}

File: Halo4API.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Runtime
Imports System.Text
Imports System.Collections.Specialized
Imports System.IO
Imports System.Net
Imports System.Text.RegularExpressions
Imports System.Web
Imports Newtonsoft.Json

Public Class Halo4API

    Private Class BetterWebRequest
        Public Shared Function Post(url As String, body As String) As HttpWebResponse  'Implies "POST"
            Dim request As HttpWebRequest = DirectCast(HttpWebRequest.Create(url), HttpWebRequest)
            request.AllowAutoRedirect = False
            request.Method = "POST"
            request.CookieContainer = Halo4API.Cookies
            Dim response As HttpWebResponse
            Dim newBody As System.Text.ASCIIEncoding = New System.Text.ASCIIEncoding()
            request.ContentLength = body.Length
            request.ContentType = "application/x-www-form-urlencoded"

            Dim writer As StreamWriter = New StreamWriter(request.GetRequestStream())
            writer.Write(body)
            writer.Close()

            response = DirectCast(request.GetResponse(), HttpWebResponse)
            response.Close()

            Return response
        End Function
    End Class

    Private Class BetterWebResponse
        Public Shared Function GetBody(response As HttpWebResponse) As String
            Try
                Dim dataStream As Stream = response.GetResponseStream()

                Dim reader As StreamReader = New StreamReader(dataStream)

                Dim body As String = reader.ReadToEnd()

                dataStream.Close()
                reader.Close()

                Return body
            Catch ex As Exception
                Return Nothing
            End Try
        End Function
    End Class

    Private Class WaypointAuthTokenDocument
        Public access_token As String
        Public AuthenticationToken As String
        Public expires_in As Integer
    End Class

    Private Class UserInformation
        Public Gamertag As String
        Public AnalyticsToken As String
    End Class

    Private Class WaypointSpartanTokenDocument
        Public ResponseCode As Integer
        Public SpartanToken As String
        Private UserInformation As UserInformation
    End Class

    'Enter an email or pass here to use the GetAPIKey() method without passing values
    Private microsoftEmail As String = ""
    Private microsoftPassword As String = ""

    Private msLogin As String = "https://login.live.com/login.srf?id=2"
    Private waypointGateway As String = "https://app.halowaypoint.com/oauth/signin?returnUrl=https%3A%2F%2Fapp.halowaypoint.com%2Fen-us"
    Private waypointRegisterUrl As String = "https://settings.svc.halowaypoint.com/RegisterClientService.svc/spartantoken/wlid?_={0}"

    'New URL's
    Private urlToScrape As String = "https://login.live.com/oauth20_authorize.srf?client_id=000000004C0BD2F1&scope=xbox.basic+xbox.offline_access&response_type=code&redirect_uri=https://www.halowaypoint.com:443/oauth/callback&state=https%3A%2F%2Fapp.halowaypoint.com%2Fen-us%2F&locale=en-US&display=touch"
    Private urlToPost As String = "https://login.live.com/ppsecure/post.srf?client_id=000000004C0BD2F1&scope=xbox.basic+xbox.offline_access&response_type=code&redirect_uri=https://www.halowaypoint.com:443/oauth/callback&state=https%3A%2F%2Fapp.halowaypoint.com%2Fen-us%2F&locale=en-US&display=touch&bk=1383096785"


    Public Shared Cookies As CookieContainer = New CookieContainer()

    Private Function GetMicrosoftLoginPPFT(Optional iterationCount As Integer = 0) As String 'Going to scrape a page and return an array of strings containing essential values for the login transaction
        Dim request As HttpWebRequest = DirectCast(HttpWebRequest.Create(urlToScrape), HttpWebRequest)
        Dim response As HttpWebResponse
        Try
            response = DirectCast(request.GetResponse(), HttpWebResponse)
        Catch ex As Exception
            response = Nothing
        End Try

        If (response Is Nothing AndAlso iterationCount < 10) Then
            Return GetMicrosoftLoginPPFT(iterationCount)
            iterationCount += 1
        ElseIf (iterationCount = 10) Then
            If (Not response Is Nothing) Then
                response.Close()
            End If
            Throw New Exception("Failed to get login parameters after 10 attempts, Microsoft may be down.")
        End If

        Dim responseBody As String = BetterWebResponse.GetBody(response)

        Dim cookieRegex As Regex = New Regex("MSPRequ=(.*?);.*,.*MSPOK=(.*?);")

        Dim cookieMatch As Match = cookieRegex.Match(response.Headers.Get("Set-Cookie"))

        Dim mspRequest As Cookie = New Cookie("MSPRequ", cookieMatch.Groups(1).Value, "/", "login.live.com")

        Dim mspOk As Cookie = New Cookie("MSPOK", cookieMatch.Groups(2).Value, cookieMatch.Groups(6).Value, "login.live.com")

        Halo4API.Cookies.Add(mspRequest)
        Halo4API.Cookies.Add(mspOk)

        response.Close()

        Dim regex As Regex = New Regex("name=""PPFT"".*?value=""(.*?)""")

        Dim match As Match = Regex.Match(responseBody)

        Dim PPFT As String

        If (match.Success) Then
            PPFT = match.Groups(1).Value
        Else
            Throw New Exception("Failed to get login parameters. Scrape returned no results, Microsoft may have changed their code.")
        End If

        Return PPFT
    End Function

    Private Function GetWaypointState(Optional iterationCount As Integer = 0) As String
        Dim request As HttpWebRequest = DirectCast(HttpWebRequest.Create(waypointGateway), HttpWebRequest)
        request.CookieContainer = Halo4API.Cookies
        Dim response As HttpWebResponse

        Try
            response = DirectCast(request.GetResponse(), HttpWebResponse)
        Catch ex As Exception
            response = Nothing
        End Try

        If (response Is Nothing AndAlso iterationCount < 10) Then
            Return GetWaypointState(iterationCount)
            iterationCount += 1
        ElseIf (iterationCount = 10) Then
            Throw New Exception("Failed to get login parameters after 10 attempts, Microsoft may be down.")
        End If

        Dim waypointState As String
        Try
            Dim queryString As NameValueCollection = System.Web.HttpUtility.ParseQueryString(response.ResponseUri.Query)
            waypointState = queryString("state")
        Catch ex As Exception
            Return GetWaypointState(++iterationCount)
        End Try

        Halo4API.Cookies.Add(response.Cookies)

        response.Close()

        Return waypointState
    End Function

    Private Function PerformWaypointLogin(email As String, pass As String) As String
        Dim PPFT As String = GetMicrosoftLoginPPFT()
        Dim PPSX As String = "Pass"

        'PPFT:   [loginPPFT]()
        'login:  [login]()
        'passwd: [passwd]()
        '           LoginOptions:3
        '           NewUser:1
        'PPSX:   [loginPPSX]()
        '           type:11
        'i3:     [rand]()
        '           m1:1680
        '           m2:1050
        '           m3:0
        '           i12:1
        '           i17:0
        '           i18:__MobileLogin|1
        Dim query As String = String.Format("PPFT={0}&login={1}&passwd={2}&LoginOptions=3&NewUser=1&PPSX={3}&type=11&i3={4}&m1=1680&m2=1050&m3=0&i12=1&i17=0&i18=__MobileLogin|1", PPFT, HttpUtility.UrlEncode(email), HttpUtility.UrlEncode(pass), PPSX, New Random().Next(15000, 50000).ToString())

        'Get the response of the query using the parameters we created
        Dim response As HttpWebResponse = BetterWebRequest.Post(urlToPost, query)

        'Add the response cookies to our container
        Halo4API.Cookies.Add(response.Cookies)

        'Close the response object
        response.Close()

        'Return the header that we need
        Return response.Headers.Get("Location")
    End Function

    Private Function GetWaypointAuthToken(homepageUrl As String) As WaypointAuthTokenDocument
        Dim request As HttpWebRequest = DirectCast(HttpWebRequest.Create(homepageUrl), HttpWebRequest)
        request.CookieContainer = Halo4API.Cookies
        request.UserAgent = "Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.17 (KHTML, like Gecko) Chrome/24.0.1312.52 Safari/537.17"

        Dim authToken As WaypointAuthTokenDocument

        Dim response As HttpWebResponse

        Try
            response = DirectCast(request.GetResponse(), HttpWebResponse)

            Dim regex As Regex = New Regex("var user.=.*?({.*?});", RegexOptions.Multiline)

            Dim body As String = BetterWebResponse.GetBody(response)

            Dim match As Match = regex.Match(body)

            If (Not match.Success) Then
                Throw New Exception()
            End If

            authToken = JsonConvert.DeserializeObject(Of WaypointAuthTokenDocument)(match.Groups(1).Value)

            Halo4API.Cookies.Add(response.Cookies)
        Catch ex As Exception
            Throw New Exception("Failed to get Waypoint auth tokens. This is likely a scraping error. Perhaps 343i changed their code layout.")
        End Try

        response.Close()

        Return authToken
    End Function

    Private Function GetWaypointSpartanToken(authToken As WaypointAuthTokenDocument) As WaypointSpartanTokenDocument
        Dim request As HttpWebRequest = DirectCast(HttpWebRequest.Create(String.Format(waypointRegisterUrl, Convert.ToInt32((DateTime.UtcNow - New DateTime(1970, 1, 1)).TotalSeconds))), HttpWebRequest)
        request.Accept = "application/json"
        request.Headers.Add("Origin", "https://app.halowaypoint.com")
        request.Headers.Add("X-343-Authorization-WLID", "v1=" + authToken.access_token)
        request.CookieContainer = Halo4API.Cookies

        Dim response As HttpWebResponse = DirectCast(request.GetResponse(), HttpWebResponse)

        Dim spartanToken As WaypointSpartanTokenDocument

        Try
            spartanToken = JsonConvert.DeserializeObject(Of WaypointSpartanTokenDocument)(BetterWebResponse.GetBody(response))
        Catch ex As Exception
            Throw New Exception("Could not parse Spartan token.")
        End Try

        Halo4API.Cookies.Add(response.Cookies)

        Return spartanToken
    End Function

    Private Function GetWaypointWebAuthTokenAndHomepage(callbackUrl As String) As String
        Dim request As HttpWebRequest = DirectCast(HttpWebRequest.Create(callbackUrl), HttpWebRequest)
        request.CookieContainer = Halo4API.Cookies
        request.AllowAutoRedirect = False

        Dim response As HttpWebResponse = DirectCast(request.GetResponse(), HttpWebResponse)

        Dim webAuth As Boolean = False

        For Each cookie As Cookie In response.Cookies
            If (cookie.Name = "WebAuth") Then
                webAuth = True
                Exit For
            End If
        Next

        If (webAuth = False) Then
            Throw New Exception("Could not get WebAuth token.")
        End If

        Halo4API.Cookies.Add(response.Cookies)

        response.Close()

        Return response.Headers.Get("Location")
    End Function

    Private Function GetSpartanTokenViaMicrosoftAuth(email As String, pass As String) As String
        Dim waypointCallbackUrl As String = PerformWaypointLogin(email, pass)
        Dim waypointHomepage As String = GetWaypointWebAuthTokenAndHomepage(waypointCallbackUrl) 'Loads the WebAuth token into the cookie store
        Dim waypointAuthToken As WaypointAuthTokenDocument = GetWaypointAuthToken(waypointHomepage)
        Dim waypointSpartanToken As WaypointSpartanTokenDocument = GetWaypointSpartanToken(waypointAuthToken)

        Return waypointSpartanToken.SpartanToken
    End Function

    Public Function GetAPIKey() As String
        Try
            Return GetSpartanTokenViaMicrosoftAuth(microsoftEmail, microsoftPassword)
        Catch e As Exception
            Throw New Exception(DirectCast(IIf(Not String.IsNullOrEmpty(e.Message), e.Message, "An unknown error occurred."), String))
        End Try
    End Function

    Public Function GetAPIKey(email As String, pass As String) As String
        Try
            Return GetSpartanTokenViaMicrosoftAuth(email, pass)
        Catch e As Exception
            Throw New Exception(DirectCast(IIf(Not String.IsNullOrEmpty(e.Message), e.Message, "An unknown error occurred."), String))
        End Try
    End Function
End Class
